### PR TITLE
Terms of Service API

### PR DESCRIPTION
**Pull request by crcommons**
This pull request makes the following changes:

- [x] Add terms of service migration to add entries when user agrees to terms of service (ToS)
- [x] Add ability to create, read, and search for ToS entries 
- [x] Searching for ToS returns the most recent entry for a given user in the db
- [x] Add tests for creating, reading, and searching for ToS Entries

Test checklist:
- [x] Run Behat Tests

Refs ushahidi/platform#1772

Ping @ushahidi/platform

### CODE REVIEW

File: migrations/20170712152806_add_tos_table.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+class AddTosTable extends AbstractMigration
+{
+    /**
+     * Migrate Up.
+     */
+    public function up()
+    {
+        $this->table('tos')
+          ->addColumn('user_id', 'integer', ['null' => false])
+          ->addColumn('agreement_date', 'integer', ['null' => false])
```
Comment (rjmackay) [inline migrations/20170712152806_add_tos_table.php]: I'd usually suggest these should be mysql datetime column, not just integers. That said... if its working as-is we should really get this deployed. This part of the codebase is unlikely to change often so not a major

File: application/classes/Ushahidi/Repository/Tos.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    protected function getTable()
+    {
+        return 'tos';
+    }
+
+
+    // CreateRepository
+    public function create(Entity $entity)
+    {
+        //get the user ID
```
Comment (rjmackay) [inline application/classes/Ushahidi/Repository/Tos.php]: Minor style issue, our comments are generally of the form `// Get user ID` ie. space after //, and Upper case start, aim for full sentences where it make sense

File: application/classes/Ushahidi/Validator/Tos/Create.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return [
+            'id' => [
+                ['numeric'],
+            ],
+            'user_id' => [
+                ['numeric'],
+                [[$this->user_repo, 'exists'], [':value']],
+            ],
+            'agreement_date' => [
+                [[$this, 'validDate'], [':value']],
```
Comment (rjmackay) [inline application/classes/Ushahidi/Validator/Tos/Create.php]: You can just use `['date']` here, similar to numeric above

File: application/classes/Ushahidi/Formatter/Tos.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * @package    Ushahidi\Application
+ * @copyright  2014 Ushahidi
+ * @license    https://www.gnu.org/licenses/agpl-3.0.html GNU Affero General Public License Version 3 (AGPL3)
+ */
+
+use Ushahidi\Core\Traits\FormatterAuthorizerMetadata;
+
+class Ushahidi_Formatter_Tos extends Ushahidi_Formatter_API
+{
+    use FormatterAuthorizerMetadata;
```
Comment (rjmackay) [inline application/classes/Ushahidi/Formatter/Tos.php]: This should have formatters for the dates - have a look at `Ushahidi_Formatter_API` for what we use for created/updated dates

File: application/config/init.php [hunk 5]
```diff
@@ -28,7 +28,7 @@
 	'base_url'    => '/',
 	'index_file'  => FALSE,
 	'charset'     => 'utf-8',
-	'errors'      => TRUE,
+	'errors'      => FALSE,
```
Comment (rjmackay) [inline application/config/init.php]: revert this before it lands?

File: src/Core/Tool/Authorizer/TosAuthorizer.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            return true;
+        }
+
+        if ($privilege === 'read') {
+            return true;
+        }
+
+
+        // If no other access checks succeed, we default to denying access
+        return true;
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/TosAuthorizer.php]: errr... this should default to `false` not `true` right?

File: src/Core/Tool/Authorizer/TosAuthorizer.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        if ($privilege === 'create') {
+            return true;
+        }
+
+        if ($privilege === 'search') {
+            return true;
+        }
+
+        if ($privilege === 'read') {
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/TosAuthorizer.php]: I think we should only allow reading when user id = current user id right?

File: src/Core/Usecase/CreateUsecase.php [hunk 8]
```diff
@@ -64,19 +64,15 @@ public function interact()
 	{
 		// Fetch a default entity and apply the payload...
 		$entity = $this->getEntity();
```
Comment (rjmackay) [inline src/Core/Usecase/CreateUsecase.php]: Can you revert the whitespace changes?

File: src/Core/Usecase/Tos/CreateTos.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ * @license    https://www.gnu.org/licenses/agpl-3.0.html GNU Affero General Public License Version 3 (AGPL3)
+ */
+
+namespace Ushahidi\Core\Usecase\Tos;
+
+use Ushahidi\Core\Entity;
+use Ushahidi\Core\Entity\Tos;
+use Ushahidi\Core\Usecase\CreateUsecase;
+
+class CreateTos extends CreateUsecase
```
Comment (rjmackay) [inline src/Core/Usecase/Tos/CreateTos.php]: You don't need to extend this if nothing changes, you can just use CreateUsecase as-is

File: src/Init.php [hunk 10]
```diff
@@ -308,6 +310,13 @@ function feature($name)
 	'create'  => $di->lazyNew('Ushahidi\Core\Usecase\Contact\CreateContact')
 ];
 
+// Add custom create usecase for terms of service
+$di->params['Ushahidi\Factory\UsecaseFactory']['map']['tos'] = [
+	'create'  => $di->lazyNew('Ushahidi\Core\Usecase\Tos\CreateTos'),
```
Comment (rjmackay) [inline src/Init.php]: This isn't needed, since you're not changing any behaviour there anyway

File: tests/integration/tos.feature [hunk 11]
```diff
@@ -0,0 +1,27 @@
+@resetFixture @tos 
+Feature: Testing the Tos API
+
+    Scenario: Create a ToS entry
+        Given that I want to make a new "tos"
+        And that the request "Authorization" header is "testbasicuser"
+        And that the request "data" is:
+            """
+            {
+                "tos_version_date":"1500059540"
```
Comment (rjmackay) [inline tests/integration/tos.feature]: We should accept and return dates in JSON date format.

File: application/classes/Ushahidi/Repository/Tos.php (hunk 2 above)
Comment (crcommons) [inline application/classes/Ushahidi/Repository/Tos.php]: For some reason the linter was getting mad with spaces (only sometimes) after the // in comments. Will try to fix and see if it works this time.

File: application/classes/Ushahidi/Validator/Tos/Create.php (hunk 3 above)
Comment (crcommons) [inline application/classes/Ushahidi/Validator/Tos/Create.php]: To discuss. I changed and broke tests.

File: application/config/init.php (hunk 5 above)
Comment (crcommons) [inline application/config/init.php]: done.

File: src/Core/Tool/Authorizer/TosAuthorizer.php (hunk 6 above)
Comment (crcommons) [inline src/Core/Tool/Authorizer/TosAuthorizer.php]: Whoops! Done.

File: src/Core/Tool/Authorizer/TosAuthorizer.php (hunk 7 above)
Comment (crcommons) [inline src/Core/Tool/Authorizer/TosAuthorizer.php]: done.

File: src/Core/Usecase/Tos/CreateTos.php (hunk 9 above)
Comment (crcommons) [inline src/Core/Usecase/Tos/CreateTos.php]: Done.

File: src/Init.php (hunk 10 above)
Comment (crcommons) [inline src/Init.php]: Done.

File: src/Core/Entity/Tos.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            'user_id' => ['user', 'user.id']
+        ];
+    }
+
+    protected function getDefinition()
+    {
+        return [
+            'id'              => 'int',
+            'user_id'         => 'int',
+            'agreement_date'  => 'int',
```
Comment (rjmackay) [inline src/Core/Entity/Tos.php]: This might need to be `'*date',`

File: src/Core/Usecase/CreateUsecase.php (hunk 8 above)
Comment (crcommons) [inline src/Core/Usecase/CreateUsecase.php]: Done, I think. Ran the command you suggested.

File: src/Core/Tool/Authorizer/TosAuthorizer.php [hunk 13]
```diff
@@ -39,9 +39,14 @@ class TosAuthorizer implements Authorizer
 
     /* Authorizer */
     public function isAllowed(Entity $entity, $privilege)
-    {
+    {   
+        // Get the current user ID
+        $current_user = service('session.user');
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/TosAuthorizer.php]: This is actually the same thing returned by `getUser` below. I think you need to compare with the $entity->user_id

**Pull request merged**